django/db/models/sql/query.py

```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
         query.promote_joins(to_promote)
         query.demote_joins(to_demote)
         return to_demote
+
+
+class ExpressionConstraintQuery:
```

Comment:
I hadn't considered that. My main focus has been trying to get something that works well for this use case. Refactoring to allow code re-use is a secondary concern.